### PR TITLE
State Trigger -> HA Form

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

Converts the State Trigger form to ha-form
Adds Duration Selector

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-base-time-input.ts [hunk 1]
```diff
@@ -115,7 +120,7 @@ export class HaBaseTimeInput extends LitElement {
           @input=${this._valueChanged}
           @focus=${this._onFocus}
           no-spinner
-          required
+          ?required=${this.required}
```
Comment (bramkragten) [inline src/components/ha-base-time-input.ts:123]:
```suggestion
          .required=${this.required}
```

File: src/components/ha-base-time-input.ts [hunk 2]
```diff
@@ -135,7 +140,7 @@ export class HaBaseTimeInput extends LitElement {
           @focus=${this._onFocus}
           name="minutes"
           no-spinner
-          required
+          ?required=${this.required}
```
Comment (bramkragten) [inline src/components/ha-base-time-input.ts:143]:
```suggestion
          .required=${this.required}
```

File: src/components/ha-base-time-input.ts [hunk 3]
```diff
@@ -156,7 +161,7 @@ export class HaBaseTimeInput extends LitElement {
               @focus=${this._onFocus}
               name="seconds"
               no-spinner
-              required
+              ?required=${this.required}
```
Comment (bramkragten) [inline src/components/ha-base-time-input.ts:164]:
```suggestion
              .required=${this.required}
```

File: src/components/ha-base-time-input.ts [hunk 4]
```diff
@@ -177,7 +182,7 @@ export class HaBaseTimeInput extends LitElement {
               @focus=${this._onFocus}
               name="milliseconds"
               no-spinner
-              required
+              ?required=${this.required}
```
Comment (bramkragten) [inline src/components/ha-base-time-input.ts:185]:
```suggestion
              .required=${this.required}
```

File: src/components/ha-selector/ha-selector-duration.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  @property({ type: Boolean }) public disabled = false;
+
+  protected render() {
+    return html`
+      <ha-duration-input
+        .label=${this.label}
+        .data=${this.value}
+        .disabled=${this.disabled}
+        .required=${!this.selector.duration.optional}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-duration.ts:25]: optional should not be part of the selector, it a level higher

File: src/components/ha-selector/ha-selector-text.ts [hunk 6]
```diff
@@ -35,7 +35,7 @@ export class HaTextSelector extends LitElement {
         autocapitalize="none"
         autocomplete="off"
         spellcheck="false"
-        required
+        ?required=${!this.selector.text.optional}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:38]:
```suggestion
        .required=${this.required}
```

File: src/components/ha-selector/ha-selector-duration.ts (hunk 5 above)
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-duration.ts:25]:
```suggestion
        .required=${this.required}
```

File: src/components/ha-selector/ha-selector-text.ts [hunk 7]
```diff
@@ -50,7 +50,7 @@ export class HaTextSelector extends LitElement {
           ? // reserve some space for the icon.
             html`<div style="width: 24px"></div>`
           : this.selector.text?.suffix}
-        required
+        ?required=${!this.selector.text.optional}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:53]:
```suggestion
        .required=${this.required}
```

File: src/data/selector.ts [hunk 8]
```diff
@@ -31,6 +31,10 @@ export interface DeviceSelector {
   };
 }
 
+export interface DurationSelector {
+  duration: { optional: boolean };
```
Comment (bramkragten) [inline src/data/selector.ts:35]:
```suggestion
  duration: {  };
```

File: src/data/selector.ts [hunk 9]
```diff
@@ -111,6 +115,7 @@ export interface StringSelector {
       | "datetime-local"
       | "color";
     suffix?: string;
+    optional?: boolean;
```
Comment (bramkragten) [inline src/data/selector.ts:118]:
```suggestion
```

File: src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts [hunk 10]
```diff
@@ -76,57 +76,35 @@ export class HaStateTrigger extends LitElement implements TriggerElement {
   }
 
   protected render() {
-    const { entity_id, attribute, to, from } = this.trigger;
-    const trgFor = createDurationData(this.trigger.for);
```
Comment (bramkragten) [inline src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts:80]: `for` should have some preprocessing

File: src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts [hunk 11]
```diff
@@ -36,6 +29,13 @@ const stateTriggerStruct = assign(
   })
 );
 
+const SCHEMA = [
+  { name: "entity_id", selector: { entity: {} } },
+  { name: "from", selector: { text: { optional: true } } },
```
Comment (bramkragten) [inline src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts:34]:
```suggestion
  { name: "from", selector: { text: { }, } },
```

File: src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts [hunk 12]
```diff
@@ -36,6 +29,13 @@ const stateTriggerStruct = assign(
...
+  { name: "to", selector: { text: { optional: true } } },
+  { name: "for", selector: { duration: { optional: true } } },
```
Comment (bramkragten) [inline src/panels/config/automation/trigger/types/ha-automation-trigger-state.ts:36]:
```suggestion
  { name: "to", selector: { text: { } } },
  { name: "for", selector: { duration: { } } },
```

File: src/components/ha-selector/ha-selector.ts [hunk 13]
```diff
@@ -31,6 +32,8 @@ export class HaSelector extends LitElement {
 
   @property({ type: Boolean }) public disabled = false;
 
+  @property() public schema?: HaFormSchema;
```
[the remark this answers is not in the document]
Comment (zsarnett) [reply src/components/ha-selector/ha-selector.ts:35]: Gotcha. Schema was already being passed so I just picked it up

[the remark this answers is not in the document]
Comment (bramkragten) [reply src/components/ha-selector/ha-selector.ts:35]: Yeah, but a selector should be usable without `ha-form`

**Review by bramkragten — APPROVED** [review]

**Pull request merged**